nsxt/resource_nsxt_upgrade_run.go

```diff
@@ -0,0 +1,568 @@
+/* Copyright © 2024 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/protocol/client"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade/plan"
+)
+
+// Order matters
+var upgradeComponentList = []string{
+	edgeUpgradeGroup,
+	hostUpgradeGroup,
+	mpUpgradeGroup,
+}
+
+var componentToGroupKey = map[string]string{
+	edgeUpgradeGroup: "edge_group",
+	hostUpgradeGroup: "host_group",
+}
+
+var componentToSettingKey = map[string]string{
+	edgeUpgradeGroup: "edge_upgrade_setting",
+	hostUpgradeGroup: "host_upgrade_setting",
+}
+
+var supportedUpgradeGroupExtendedConfigValue = map[string][]string{
+	"upgrade_mode":                                     {"maintenance_mode", "in_place", "stage_in_vlcm"},
+	"maintenance_mode_config_vsan_mode":                {"evacuate_all_data", "ensure_object_accessibility", "no_action"},
+	"maintenance_mode_config_evacuate_powered_off_vms": {"true", "false"},
+	"rebootless_upgrade":                               {"true", "false"},
+}
+
+var (
+	upgradeStatusCheckInterval = 2 * time.Minute
```

Comment:
should we make this configurable as part of the resource?

Reply:
Done.